checker/env.go

```diff
@@ -57,20 +58,49 @@ func (e *Env) Add(decls ...*checked.Decl) {
 	}
 }
 
-func (e *Env) addFunction(decl *checked.Decl) {
-	current := e.declarations.FindFunction(decl.Name)
-	if current != nil {
-		if current.Name != decl.Name {
+func (e *Env) addOverload(f *checked.Decl, overload *checked.Decl_FunctionDecl_Overload) {
+	function := f.GetFunction()
+	emptyMappings := types.NewMapping()
+	overloadFunction := types.NewFunction(overload.GetResultType(),
+		overload.GetParams()...)
+	overloadErased := types.Substitute(emptyMappings, overloadFunction, true)
+	for _, existing := range function.GetOverloads() {
+		existingFunction := types.NewFunction(existing.GetResultType(),
+			existing.GetParams()...)
+		existingErased := types.Substitute(emptyMappings, existingFunction, true)
+		overlap := (types.IsAssignable(emptyMappings, overloadErased, existingErased) != nil ||
+			types.IsAssignable(emptyMappings, existingErased, overloadErased) != nil)
+		if overlap &&
+			overload.GetIsInstanceFunction() == existing.GetIsInstanceFunction() {
+			e.errors.overlappingOverload(common.NoLocation, f.Name, overloadFunction,
+				existingFunction)
+			return
+		}
+	}
+
+	for _, macro := range parser.AllMacros {
```

Comment:
We should consider moving the Macros into common/macros/macros.go. Operators and overloads are currently in their own packages there.

Reply:
I did look into this and did not do it. I thought this would be useful if we could do macros construction only in common/macros/macros.go and expose AllMacros only(similar to what you did with operators and overloads). But macros initialization seem to be tightly integrated with parser because of expander. I could make macros.go use an interface for parserHelper but that did not seem very appealing.

Reply:
There are portions of the `parserHelper` that should be made an interface, specifically expression id / source location tracking, but this can wait. Seems like making the macro metadata common and the implementations specific to the parser would be the ideal way to go. I would prefer to not have the checker depend directly on the parser if possible, but let's just add a TODO for now.

Reply:
Added a TODO.